document `cargoLock.lockFileContents`

dermetfan opened the pull request:
###### Motivation for this change

As requested in https://github.com/NixOS/nixpkgs/pull/137395#pullrequestreview-758827443

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

dermetfan committed:
doc: rust: document cargoLock.lockFileContents

dermetfan commented on doc/languages-frameworks/rust.section.md:
```diff
@@ -122,16 +122,38 @@ rustPlatform.buildRustPackage rec {
   cargoLock = {
     lockFile = ./Cargo.lock;
-  }
+  };
```
Not related but also not worth another PR

figsoda commented on doc/languages-frameworks/rust.section.md:
```diff
@@ -122,16 +122,38 @@ rustPlatform.buildRustPackage rec {
+  version = "1.0.0";
+
+  cargoLock = let
+    fixupLockFile = path: /* ... */;
```
maybe this will be more clear that its a string?
```suggestion
    fixupLockFile = path: f (builtins.readFile path);
```

dermetfan committed:
doc: rust: improve clarity of example snippet

Make it more clear that the return type is a string.

Co-authored-by: figsoda <figsoda@pm.me>

happysalada commented on doc/languages-frameworks/rust.section.md:
```diff
@@ -122,16 +122,38 @@ rustPlatform.buildRustPackage rec {
+the specified lockfile.
+
+Alternatively, `cargoLock.lockFileContents` can be set to a string of
+the contents of a `Cargo.lock` file, for example if you need to
```
I would add in brackets after file something like
`(in case the src Cargo.lock needs patching. This cannot be done in patches, since the patches run after the fetching of the dependencies)`

dermetfan replied:
Good point. I rephrased the entire paragraph to hopefully convey the purpose more clearly.

dermetfan committed:
doc: rust: simplify snippet

Do not use recursive attribute sets where not needed, it could confuse the reader.

dermetfan committed:
doc: rust: rephrase paragraph about `cargoLock.lockFileContents`

mentioning that `Cargo.lock` cannot be patched in the `patchPhase`

dermetfan commented on doc/languages-frameworks/rust.section.md:
````diff
@@ -116,22 +116,44 @@ is updated after every change to `Cargo.lock`. Therefore,
 
 ```nix
-rustPlatform.buildRustPackage rec {
+rustPlatform.buildRustPackage {
````
`rec` is not needed here and could confuse the reader. Not really related but I let's keep this snippet and the new, very similar one below in sync.

happysalada commented:
Thank you for this!
